### PR TITLE
fix(asr): warm up Numba RNNT and TDT losses before training

### PR DESCRIPTION
**Pull request by albertz**
# What does this PR do ?

Warm up Numba RNNT and TDT losses before training to reduce memory retained by compilation.

**Collection**: ASR

# Changelog

- Automatically warm supported loss kernels at training startup, including both TDT and RNNT branches.
- Use small inputs, preserve random states, and collect compiler reference cycles before training activations exist.
- Add reusable warmup methods, tests, and documentation.

# Usage

Automatic for supported Numba losses on CUDA; no configuration changes are needed.
Custom training loops can call `loss.warmup(device)` before allocating batch activations.

# Before your PR is "Ready for review"

**Pre checks**:

- [x] Read and followed the contributor guidelines.
- [x] Added necessary tests.
- [x] Updated documentation.
- [x] Affects an optional component: Numba.
  - [ ] Reviewer: Verify import guards for optional libraries.

**PR Type**:

- [ ] New Feature
- [x] Bugfix
- [ ] Documentation

# Additional Information

Validation:
- Black 24.10.0, isort 5.13.2, and whitespace checks passed.
- 16 isolated CPU checks passed using extracted source and mocked loss kernels.
- Four CUDA checks and the full NeMo model-fixture tests remain unrun.

### CODE REVIEW

**Review by artbataev — CHANGES_REQUESTED** [review]
The idea looks good, but the implementation seems too complex for such simple logic (taking partial, passing self).
My suggestion: use `warmup` method for loss to implement actual logic of `_warmup_tdt_loss`, `_warmup_rnnt_loss`. To avoid duplication of the code from `_warmup_numba_loss`, convert it to a separate context manager.

**Comment by albertz** [comment]
The context: I found some bugs in Numba, Numba-CUDA, Numba-CUDA-MLIR, that are the reason for why I got OOM in Nemo Speech. It's somewhat obfuscated. Numba stores exceptions in some object, some `failures` tracker. The exceptions contain the frames and thus reference the local variables, even of parent caller frames. That `failures` object goes then out of scope when the func returns, but that introduces a reference cycle (frame object -> `failures` obj -> exception -> traceback -> frame object), and this is not cleaned by the normal ref counter, but much later by some next GC cycle. The parent frames keep references to some previous activations. So because of the reference cycle and the late deletion, it is still in GPU mem, and that causes the CUDA OOM.

**Review by artbataev — DISMISSED** [review]
LGTM, thanks!

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py [hunk 1]
```diff
@@ -411,6 +416,33 @@ def __init__(self, blank=0, reduction='mean', fastemit_lambda: float = 0.0, clam
...
+        Preserves random states and the configured loss; returns False on CPU.
+        """
+        device = torch.device(device)
+        if device.type != 'cuda':
+            return False
+
+        def warmup():
+            # Non-singleton dimensions preserve the contiguous production array layout.
+            acts = torch.zeros(
+                (2, 8, 4, max(2, self.blank + 1)), device=device, dtype=torch.float32, requires_grad=True
```
Comment (nithinraok) [inline nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py]: Does this warmup also need to cover FP16 for standard RNNT?

`RNNTLoss.forward()` can pass FP16 inputs to `RNNTLossNumba` when Numba FP16 support is available. Since Numba specializes CUDA kernels by input dtype, the FP32 warmup might not prepare the FP16 specialization.

Could we do something like this, with `RNNTLossNumba.warmup()` updated to accept `dtypes`?

```python
if not NUMBA_RNNT_AVAILABLE:
    return False

if isinstance(self._loss, RNNTLossNumba):
    dtypes = [torch.float32]
    if not self._force_float32 and numba_utils.is_numba_cuda_fp16_supported():
        dtypes.append(torch.float16)
    return self._loss.warmup(device, dtypes=dtypes)

if isinstance(self._loss, TDTLossNumba):
    return self._loss.warmup(device)

return False
```

**Review by nithinraok — COMMENTED** [review]
One question about FP16 warmup coverage.

File: nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py (hunk 1 above)
Comment (albertz) [reply nemo/collections/asr/parts/numba/rnnt_loss/rnnt_pytorch.py]: Added that.

**Review by artbataev — APPROVED** [review]
